crates/client/data-availability/src/utils.rs

```diff
@@ -146,3 +159,193 @@ pub fn bytes_to_felt(raw: &[u8]) -> StarkFelt {
 pub fn bytes_to_key(raw: &[u8]) -> PatriciaKey {
     PatriciaKey(bytes_to_felt(raw))
 }
+
+/// Width of the field storing the number of storage updates in `U256`.
+#[allow(dead_code)]
+const NUM_STORAGE_UPDATES_WIDTH: u64 = 64; // Adjust this based on your logic
+
+/// Macro for converting a value of type `U256` to a StarkNet type.
+///
+/// Usually in starknet, some types like ClassHash and ContractClass can be directly converted from
+/// Felt252Wrapper, so we use this macro to convert U256 into Felt252Wrapper and then into the
+/// corresponding data.
+///
+/// # Arguments
+///
+/// * `$data` - The value to be converted.
+/// * `$target_type` - The StarkNet type to convert to.
+///
+/// # Returns
+///
+/// A `Result` containing the converted value or an `Error` if the conversion fails.
+///
+/// # Example
+///
+/// ```rust
+/// // Example usage
+/// let result = convert_to_starknet_type!(U256::from(1), ClassHash);
+/// ```
+macro_rules! convert_to_starknet_type {
+    ($data:expr, $target_type:ident) => {{
+        let result = Felt252Wrapper::try_from($data).map(|ft| $target_type::from(ft));
+        result
+    }};
+}
+
+/// Checks if a contract is deployed based on its address and block hash.
+///
+/// # Arguments
+///
+/// * `address` - Address of the contract.
+/// * `block_hash` - Hash of the block in which the contract is being checked.
+/// * `client` - StarkNet runtime client.
+///
+/// # Returns
+///
+/// A `Result` indicating whether the contract is deployed or not.
+///
+/// # Example
+///
+/// ```rust
+/// // Example usage
+/// let result = contract_deployed(address, block_hash, client);
+/// ```
+#[allow(unused)]
+fn contract_deployed<B, C>(address: ContractAddress, block_hash: B::Hash, client: Arc<C>) -> Result<bool, Error>
+where
+    B: BlockT,
+    C: ProvideRuntimeApi<B> + HeaderBackend<B>,
+    C::Api: StarknetRuntimeApi<B>,
+{
+    #[cfg(test)]
+    // When testing, we have only an empty client.
+    return Ok(false);
+
+    #[cfg(not(test))]
+    match client.runtime_api().contract_class_hash_by_address(block_hash, address) {
+        Ok(class_hash) => Ok(!class_hash.eq(&ClassHash::default())),
+        Err(e) => Ok(false),
+    }
+}
+
+/// Decodes a state difference using the starknet v0.11.0 logic.
+///
+/// # Arguments
+///
+/// * `encoded_diff` - Encoded state difference data in the form of `U256` slices.
+/// * `block_hash` - Hash of the block in which the state difference occurred.
+/// * `client` - StarkNet runtime client.
+///
+/// # Returns
+///
+/// A `Result` containing the decoded `StateDiff` or an `Error` if decoding fails.
+///
+/// # Example
+///
+/// ```rust
+/// // Example usage
+/// let result = decode_011_diff(&encoded_diff, block_hash, client);
+/// ```
+pub fn decode_011_diff<B, C>(encoded_diff: &[U256], block_hash: B::Hash, client: Arc<C>) -> Result<StateDiff, Error>
```

Comment:
This whole method code is really hard to read.
Can you add a link to what is the `starknet v0.11.0 logic`?
And some comments in the impl itself.

I have no idea what is going on rn

Reply:
This method basically does the opposite than the `block_data_to_calldata` method.

Reply:
Also added more comments